dpctl/tensor/_dlpack.pyx

```diff
@@ -1005,31 +1005,60 @@ def from_dlpack(x, /, *, device=None, copy=None):
             X = dpt.from_dlpack(C)
 
     """
-    if not hasattr(x, "__dlpack__"):
+    dlpack_attr = getattr(x, "__dlpack__", None)
+    dlpack_dev_attr = getattr(x, "__dlpack_device__", None)
+    if not callable(dlpack_attr) or not callable(dlpack_dev_attr):
         raise TypeError(
             f"The argument of type {type(x)} does not implement "
-            "`__dlpack__` method."
-        )
-    dlpack_attr = getattr(x, "__dlpack__")
-    if not callable(dlpack_attr):
-        raise TypeError(
-            f"The argument of type {type(x)} does not implement "
-            "`__dlpack__` method."
+            "`__dlpack__` and `__dlpack_device__` methods."
         )
     try:
         # device is converted to a dlpack_device if necessary
         dl_device = None
         if device:
             if isinstance(device, tuple):
                 dl_device = device
+                if len(dl_device) != 2:
+                    raise ValueError(
+                        "Argument `device` specified as a tuple must have length 2"
+                    )
             else:
                 if not isinstance(device, dpctl.SyclDevice):
                     d = Device.create_device(device).sycl_device
-                    dl_device = (device_OneAPI, get_parent_device_ordinal_id(<c_dpctl.SyclDevice>d))
                 else:
-                    dl_device = (device_OneAPI, get_parent_device_ordinal_id(<c_dpctl.SyclDevice>device))
+                    d = device
+                dl_device = (device_OneAPI, get_parent_device_ordinal_id(<c_dpctl.SyclDevice>d))
         dlpack_capsule = dlpack_attr(max_version=get_build_dlpack_version(), dl_device=dl_device, copy=copy)
         return from_dlpack_capsule(dlpack_capsule)
     except TypeError:
```

Comment:
The intention of this exception (per array API) is that if `max_version` keyword isn't supported, a TypeError is thrown, and the legacy path is taken. Leaning on this TypeError to move a NumPy array to kDLOneAPI device could somewhat undermine that.

We will need either another try-except block or a check for the `max_version` keyword to remain conformant.

Reply:
Maybe the best way to do this would be to capture the exception in a local variable and check if the exception is specifically `__dlpack__() got an unexpected keyword argument 'max_version'` then take the legacy path, otherwise do this.

Reply:
Yes, agreed. The `dl_device` has been added at the same time as the support for `max_version` though. 

Perhaps we should expect `__dlpack__` call to raise `BufferError` if the requested `dl_device` could not be supported and trigger copying via host logic.

Reply:
> Yes, agreed. The `dl_device` has been added at the same time as the support for `max_version` though.
> 
> Perhaps we should expect `__dlpack__` call to raise `BufferError` if the requested `dl_device` could not be supported and trigger copying via host logic.

This would be a good solution, though we would still have to be careful with the source of the exception, as `copy` keyword also throws a `BufferError` when it can't be honored.